impl : broadcast multi-messages #97

GunHee5719 opened the pull request:
## Related Issue

resolve: #97 

## Description

- implements BroadcastMessages() at conn.go that broadcast multi-messages to all nodes one by one.
- implements test case for BroadcastMessages at `conn_test.go`
- add one more case for TestConnectionPool_BroadcastMessage at `conn_test.go`
- modify setting port at `conn_test.go` and `comm_test.go`

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] Applying goimports
- [x] Test case
- [x] End of Work

GunHee5719 committed:
impl : broadcast multi-messages #97

GunHee5719 commented:
I need idea for function naming. I think BroadcastMessages() make confuse with BroadcastMessage()

zeroFruit commented:
`BroadcastMultiMessages`?

GunHee5719 commented:
Not bad.. But @AgwaB rejected it

zeroFruit commented:
What's your opinion? @AgwaB

zeroFruit commented:
or how about declaring 

```
type Broadcaster interface {
    SendMessage
}

type MultiMsgBroadcaster interface {
    SendMessages
}
```
To clarify the responsibility.

ref: https://core.ac.uk/download/pdf/82733544.pdf

GunHee5719 commented:
I think it's good way that divide interface, but also i think SendMessage may make confuse with Send()

AgwaB commented:
I suggest
```go
type Sharer interface {
       Broadcast()
}

type Distributer interface {
       Broadcast()
}
```

GunHee5719 commented:
Then how about

GunHee5719 commented:
Then how about combining your opinions like this?
```go
type Sharer interface {
      BroadcastSingleMsg()
}

type Distributer interface {
      BroadcastMultipleMsg()
}
```

zeroFruit commented:
@GunHee5719 @AgwaB 
how about unify interface ?
```
type Broadcaster interface {
    Broadcast(msgList []pb.Message)
}
```
what `Broadcaster` should do is **just sending one message from list to its one connection.**

we don't care whether `msgList`'s message is same or not. just sending one message to one connection.

if messages are same then it is single message broadcasting, if not, multi message broadcasting

hihiboss commented on conn_test.go:
```diff
@@ -195,12 +210,133 @@ func TestConnectionPool_Broadcast(t *testing.T) {
+func TestConnectionPool_BroadcastMessages(t *testing.T) {
+	cp := cleisthenes.NewConnectionPool()
+
+	cAddress1 := cleisthenes.Address{"127.0.0.1", 8080}
```
Port `8080` is often used. For example, tomcat(web server) uses `8080` as its default port.
So, I recommend to change the port number :)

GunHee5719 commented:
I added code for getting available port and modified the part for setting port at whole `conn_test.go` and `comm_test.go`.

So now I think we need to decide what func name to use

AgwaB commented on conn_test.go:
```diff
@@ -135,29 +151,33 @@ func TestGrpcConnection_Close(t *testing.T) {
-		t.Fatal(err2)
+	mockStreamWrapperList := make([]*mock.StreamWrapper, 3)
+	connList := make([]cleisthenes.Connection, 3)
+	var err error
```
what is it?

GunHee5719 replied:
I made it because I need error variable for get return value at 168 line.

GunHee5719 commented:
I divided the interface as Sharer & Distributer and pushed. But because of this work, error occurred at RBC struct in `rbc.go` because there are variable `broadCaster cleisthenes.Broadcaster`. Then how can I fix this? Please select options.
 - Make new variable at `rbc.go`
 - Do not divide the interface at `conn.go`
 - etc..

And please give me some advice at naming..

zeroFruit commented on conn.go:
```diff
@@ -179,8 +179,14 @@ func (conn *GrpcConnection) readStream(errChan chan error) {
 
-type Broadcaster interface {
-	Broadcast(msg pb.Message)
+type Sharer interface {
```
I think it is better to integrate `Sharer`, `Distributor` into `MessageBroadcaster`

zeroFruit commented on conn_test.go:
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
+}
+
+func Execute_BroadcastMultipleMessage(textList []string, waitSize int) ([]bool, error) {
+	cp := cleisthenes.NewConnectionPool()
```
### change function name into camel-case & private function

### refactor large function

I think what this function do is 

1. make connection list
2. register mock handler for each connection
3. ...

I think we can make function prettier & clear

zeroFruit commented on conn_test.go:
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
+	var err error
+	err = nil
+
+	for i, _ := range connList {
```
`for i, _` => `for i`

zeroFruit commented on conn_test.go:
```diff
@@ -195,14 +212,152 @@ func TestConnectionPool_Broadcast(t *testing.T) {
+	var wg sync.WaitGroup
+	wg.Add(waitSize)
+
+	checkReceive := make([]bool, len(textList))
```
msgReceivedChecker, msgReceivedCheckList? anyway It would be better variable name ends with *noun*

zeroFruit commented:
test failed

zeroFruit reviewed (approved):
LGTM
